### ISSUE TITLE
A misdiagnosed case of type mismatch

### ISSUE THREAD

**Issue by XFFXFF** [rust-analyzer/rust-analyzer#11847]
**rust-analyzer version**: bc08b8eff 2022-03-28 stable

**rustc version**: 1.59.0 (9d1b2106e 2022-02-23)

![image](https://user-images.githubusercontent.com/30254428/160796054-35bed653-d3b3-4e17-8b35-e989ac8603a9.png)
I tried the example in https://smallcultfollowing.com/babysteps//blog/2022/03/29/dyn-can-we-make-dyn-sized/ and found that r-a gives wrong diagnostics.  

The source code:  
```rust
use std::fmt::Debug;

fn print_me_later(x: &dyn Debug) -> impl FnOnce() + '_ {
    move || println!("{x:?}")
}

fn main() {
    let f = print_me_later(&22);
    f()
}
```

**Comment by flodiebold** [comment]
Interestingly, it works with `&22i32`, so it seems to be some problem with the handling of integer variables.

**Comment by bitgaoshu** [comment]
the literal, like 22, is actually an ```InferenceVar``` , except assign a type. In inferring , the ```InferenceVar``` with kind ```Integer``` can not coercion to Debug, but can coerce to Clone, Sized etc. When infer ended, some inferenceVar will fallback to default. So, hover on 22 is i32.  Maybe what we need is to check coerce again after the fallback to eliminate mismatch

**Comment by flodiebold** [comment]
Hmm, I don't think that's how rustc handles this. I _think_ rustc doesn't even try to prove that the variable implements `Debug` while trying to decide whether to coerce, and just records that as an obligation for later. We only coerce if we can prove either uniquely or with definite guidance that the coercion is possible:
https://github.com/rust-lang/rust-analyzer/blob/1f709d54463972b189a3120be4073c507f2fbc00/crates/hir-ty/src/infer/coerce.rs#L613-L618
I think if we accepted any ambiguous result like the FIXME says this might work, though I'm not sure whether it would cause other problems (this requires at least testing on some codebases that it doesn't cause regressions). rustc does this in a more sophisticated way (it has its own trait solving loop specially for coercion), and I'm not sure whether this has any functional differences or is just for better diagnostics.

**Comment by flodiebold** [comment]
I don't think that's a good approach to fix this.

**Comment by bitgaoshu** [comment]
yes. 
1.  we have no enough info to decide what the literal type is. the hover i32 is a fallback, a result from L432. In L432, it just covert a  ```InferenceVar, kind = Integer``` to I32, and no check overflow. so like , Issue #12155.  Maybe it's on purpose. 
2. If we change ```Debug``` to ```Clone```, of course not runnable, the mismatch info disappeared.  Clone is builtin of ```InferenceVar, kind = Integer```.  but ```Debug``` is not.
<img width="791" alt="截屏2022-05-05 12 03 09" src="https://user-images.githubusercontent.com/7943146/166862163-69a99100-57b0-4a52-be20-21c2d3fb7766.png">


So I think. there are two ways to correctly record the coercion. one is use default i32, the other is  ```InferenceVar, kind = Integer```

**Comment by flodiebold** [comment]
This problem isn't really related to fallback, it can also happen if we later do find out the concrete type for the literal:
```rust
fn main() {
    let i = 22;
    let f = print_me_later(&i);
    let u: u64 = i;
    f()
}
```
The thing that matters is that it's a variable at the point where we need to coerce, and we can't go back later and change the decision.

**Comment by bitgaoshu** [comment]
The mismatch's  expected  is decided. and the actual is not finally decided =>  When the variable's type  is finally decided, the mismatch is finally decided.

**Comment by flodiebold** [comment]
As I mentioned before, it's not just about the type mismatch, we need to decide the coercion correctly. Here's a more complicated example where this matters for type inference down the line:
```rust
trait Foo<T> { fn get(&self) -> T; }

impl Foo<&'static str> for i32 {
    fn get(&self) -> &'static str {
        "ok"
    }
}

impl Foo<u32> for u32 {
    fn get(&self) -> u32 { *self }
}

fn test() {
    let x = 2;
    let d: &dyn Foo<_> = &x;
    let _: i32 = x;

    eprintln!("{}", d.get().len());
}
```

**Comment by bitgaoshu** [comment]
oh. Just for a diagnostic.  For this issue, no matter what the literal type is , the diagnostic should not show. So, I think,  just check if some literal type matches.

**Comment by bitgaoshu** [comment]
I also test  ```// FIXME actually we maybe should also accept unknown guidance here ```.  for now, it accept all test

**Comment by WaffleLapkin** [comment]
This is marked as https://github.com/rust-lang/rust-analyzer/labels/S-actionable, but is it really? Does anyone know how to fix r-a's type inference?

**Comment by flodiebold** [comment]
Yes, this is probably blocked on the new trait solver.

**Comment by lasantosr** [comment]
> Yes, this is probably blocked on the new trait solver.

Is it unblocked now that the new trait solver has stabilized in 1.84?

**Comment by WaffleLapkin** [comment]
@lasantosr 1.84 only stabilized the use of the new trait solver _in coherence_, i.e. it is being used to check that trait `impl`s don't overlap. Stabilization of the new trait solver is also mostly independent from work on using the new trait solver in r-a -- the trait solver is "just" a library, the fact that rustc is using it does not necessarily affect r-a.

If you want to follow integration of r-a and the new trait solver there is this zulip thread:

**Comment by lnicola** [comment]
Fixed in the new trait solver.
